aten/src/ATen/native/cuda/ReduceSumProdKernel.cu

```diff
@@ -7,61 +7,100 @@
 
 namespace at { namespace native {
 
-template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
-void sum_kernel_impl(TensorIterator& iter) {
-  gpu_reduce_kernel<scalar_t, out_t>(iter, func_wrapper<out_t> ([]GPU_LAMBDA(acc_t a, acc_t b) -> acc_t {
-    return a + b;
-  }));
-}
+template <typename scalar_t, typename acc_t = scalar_t, typename out_t = scalar_t>
+struct sum_functor {
+  void operator()(TensorIterator& iter) {
+    gpu_reduce_kernel<scalar_t, out_t>(
+        iter, func_wrapper<out_t>([] GPU_LAMBDA(acc_t a, acc_t b) -> acc_t {
+          return a + b;
+        }));
+  }
+};
 
-template <typename scalar_t, typename acc_t=scalar_t, typename out_t=scalar_t>
-void prod_kernel_impl(TensorIterator& iter) {
-  gpu_reduce_kernel<scalar_t, out_t>(iter, func_wrapper<out_t> ([]GPU_LAMBDA(acc_t a, acc_t b) -> acc_t {
-    return a * b;
-  }), 1);
-}
+template <typename scalar_t, typename acc_t = scalar_t, typename out_t = scalar_t>
+struct nansum_functor {
+  void operator()(TensorIterator& iter) {
+    gpu_reduce_kernel<scalar_t, out_t>(
+        iter, NanSumOps<acc_t, out_t>{});
+  }
+};
+
+template <typename scalar_t, typename acc_t = scalar_t, typename out_t = scalar_t>
+struct prod_functor {
+  void operator()(TensorIterator& iter) {
+    gpu_reduce_kernel<scalar_t, out_t>(
+        iter, func_wrapper<out_t>([] GPU_LAMBDA(acc_t a, acc_t b) -> acc_t {
+          return a * b;
+        }), 1);
+  }
+};
 
-static void sum_kernel_cuda(TensorIterator& iter) {
+// The function `reduce_dispatch` below dispatches to the kernel based
+// on the type of `iter`. It takes care of the common logic
+// for handling Half-Precision floating types.
+// Otherwise the functor `op` is called to dispatch to the kernel
+// of relevant type.
+//
+// Note: Functor `op` should take care of all the types to be supported
+//       except for `at::Half` and `at::BFloat16`.
+template <
+    template <
+        typename scalar_t,
+        typename acc_t = scalar_t,
+        typename out_t = scalar_t>
+    typename OpFunctor,
+    typename GeneralDispatcher>
+static void reduce_dispatch(TensorIterator& iter, GeneralDispatcher op) {
```

Comment:
The code here looks elegant, but is there a performance impact on the existing functions, like sum, by building the callable structs each time this function is called?

Reply:
Follow-up question from look the the templates: the functions of interest all have the same signature, right? 

Could this be written, for example, as one function for each op that called a helper that handled the common part (lines 54-67 below) and then implemented the function-specific dispatch? Further, can the helper avoid be a function template by using the common signature of these functions to specify its function pointer argument?

Reply:
I think it should not be an issue as the structs method will get inlined and construction of that struct is trivial.
Reference: https://stackoverflow.com/a/18753022/5602957

Also tried simulating a similar code on CompilerExplorer
https://godbolt.org/z/7WTWhe

Compiler is able to actually deduce the finally value, so it don't think this structure should hinder the compiler optimizations.

> Follow-up question from look the the templates: the functions of interest all have the same signature, right?
> 
> Could this be written, for example, as one function for each op that called a helper that handled the common part (lines 54-67 below) and then implemented the function-specific dispatch? Further, can the helper avoid be a function template by using the common signature of these functions to specify its function pointer argument?

Slightly confused. Could you please give sample code.

Thanks for looking into it.

Reply:
> I think it should not be an issue as the structs method will get inlined and construction of that struct is trivial.
> Reference: https://stackoverflow.com/a/18753022/5602957
> 
> Also tried simulating a similar code on CompilerExplorer
> https://godbolt.org/z/7WTWhe
> 
> Compiler is able to actually deduce the finally value, so it don't think this structure should hinder the compiler optimizations.
> 
> > Follow-up question from look the the templates: the functions of interest all have the same signature, right?
> > Could this be written, for example, as one function for each op that called a helper that handled the common part (lines 54-67 below) and then implemented the function-specific dispatch? Further, can the helper avoid be a function template by using the common signature of these functions to specify its function pointer argument?
> 
> Slightly confused. Could you please give sample code.
> 
> Thanks for looking into it.

Thanks for investigating. I think that addresses my concern so this should be fine.